xds/src/main/java/io/grpc/xds/XdsClientImpl.java

```diff
@@ -0,0 +1,725 @@
+    // corresponds to.
+    // A nonce becomes stale following a newer nonce being presented to the client in a
+    // DiscoveryResponse.
+    private String ldsRespNonce = "";
+    private String rdsRespNonce = "";
+
+    // Most recently requested resource name(s) for each resource type. Note the resource_name in
+    // LDS requests will always be "xds:" URI (including port suffix if present).
+    @Nullable
+    private String rdsResourceName;
+
+    private AdsStream(AggregatedDiscoveryServiceGrpc.AggregatedDiscoveryServiceStub stub) {
+      this.stub = checkNotNull(stub, "stub");
+    }
+
+    private void start() {
+      requestWriter = stub.withWaitForReady().streamAggregatedResources(this);
+    }
+
+    @Override
+    public void onNext(final DiscoveryResponse response) {
+      syncContext.execute(new Runnable() {
+        @Override
+        public void run() {
+          responseReceived = true;
+          String typeUrl = response.getTypeUrl();
+          if (typeUrl.equals(ADS_TYPE_URL_LDS)) {
+            handleLdsResponse(response);
+          } else if (typeUrl.equals(ADS_TYPE_URL_RDS)) {
+            handleRdsResponse(response);
+          }
+          // TODO(zdapeng): add CDS/EDS response handles.
+        }
+      });
+    }
+
+    @Override
+    public void onError(final Throwable t) {
+      syncContext.execute(new Runnable() {
+        @Override
+        public void run() {
+          handleStreamClosed(
+              Status.fromThrowable(t).augmentDescription("ADS stream [" + this + "] had an error"));
+        }
+      });
+    }
+
+    @Override
+    public void onCompleted() {
+      syncContext.execute(new Runnable() {
+        @Override
+        public void run() {
+          handleStreamClosed(
+              Status.UNAVAILABLE.withDescription("ADS stream [" + this + "] was closed"));
+        }
+      });
+    }
+
+    private void handleStreamClosed(Status error) {
+      checkArgument(!error.isOk(), "unexpected OK status");
+      if (closed) {
+        return;
+      }
+      closed = true;
+      cleanUp();
+      configWatcher.onError(error);
+      if (responseReceived || retryBackoffPolicy == null) {
+        // Reset the backoff sequence if had received a response, or backoff sequence
+        // has never been initialized.
+        retryBackoffPolicy = backoffPolicyProvider.get();
+      }
+      long delayNanos = 0;
+      if (!responseReceived) {
+        delayNanos =
+            Math.max(
+                0,
+                retryBackoffPolicy.nextBackoffNanos() - stopwatch.elapsed(TimeUnit.NANOSECONDS));
+      }
+      logger.log(Level.FINE, "{0} stream closed, retry in {1} ns", new Object[]{this, delayNanos});
+      if (delayNanos == 0) {
+        startRpcStream();
+        if (configWatcher != null) {
+          adsStream.sendXdsRequest(ADS_TYPE_URL_LDS, ImmutableList.of(ldsResourceName));
+        }
+        // TODO(chengyuanzhang): send CDS/EDS requests if CDS/EDS watcher presents.
```

Comment:
You can schedule `RpcRetryTask` even `delayNanos == 0`, then you don't need these //TODO. 

The test may break, but can be easily fixed by just drain the pending tasks. 

Scheduling task with zero delay instead of executing inline can avoid bugs if there were some more inline things happening after `handleStreamClosed()`

Reply:
Changed to schedule a 0-delay retry task for immediate retries.

It's a bit awkward to explicitly call `fakeClock.runDueTasks()` for immediate retries in test... (I would prefer this happens itself naturally). But, whatever...